common/util/tree_operations.h

```diff
@@ -0,0 +1,926 @@
+// An only-child, first-child, and a tree root will have birth rank 0.
+// This function is aware of children container type: it uses `std::distance()`
+// to calculate node index in random access containers (which usually performs
+// in O(1)), and O(n) loop for other containers.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::Parent::available>* = nullptr>
+size_t BirthRank(const T& node) {
+  using Iterator = decltype(node.Parent()->Children().begin());
+  return tree_operations_internal::BirthRank(
+      node, typename std::iterator_traits<Iterator>::iterator_category());
+}
+
+// Returns the number of parents between this node and the root.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::Parent::available>* = nullptr>
+size_t NumAncestors(const T& node) {
+  size_t depth = 0;
+  for (const T* p = node.Parent(); p != nullptr; p = p->Parent()) {
+    ++depth;
+  }
+  return depth;
+}
+
+// Returns true if 'other' is an ancestor of this node, in other words,
+// this node is descended from 'other'.
+// This method could have been named IsDescendedFrom().
+// nullptr is never considered an ancestor of any node.
+// 'this' node is not considered an ancestor of itself.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::Parent::available>* = nullptr>
+bool HasAncestor(const T& node, const T* other) {
+  if (other == nullptr) return false;
+  for (const auto* p = node.Parent(); p != nullptr; p = p->Parent()) {
+    if (p == other) return true;
+  }
+  return false;
+}
+
+// Overload for nullptr. Always returns false.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::Parent::available>* = nullptr>
+constexpr bool HasAncestor(const T& node, std::nullptr_t) {
+  return false;
+}
+
+// Returns reference to the tree root, the greatest ancestor of this node.
+// The const-ness of the returned reference matches the const-ness of the node
+// argumnent.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::Parent::available>* = nullptr>
+T& Root(T& node) {
+  T* root = &node;
+  while (root->Parent() != nullptr) {
+    root = root->Parent();
+  }
+  return *root;
+}
+
+// Returns the closest common ancestor to this and the other, else nullptr.
+// The const-ness of the returned pointer matches the const-ness of the node
+// argumnent.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::Parent::available>* = nullptr>
+T* NearestCommonAncestor(T& node_a, T& node_b) {
+  T* a = &node_a;
+  T* b = &node_b;
+  std::set<T*> ancestors_a, ancestors_b;
+  // In alternation, insert a/b into its respective set of ancestors,
+  // and check for membership in the other ancestor set.
+  // Return as soon as one is found in the other's set of ancestors.
+  while (a != nullptr || b != nullptr) {
+    if (a != nullptr) {
+      if (ancestors_b.find(a) != ancestors_b.end()) {
+        return a;
+      }
+      ancestors_a.insert(a);
```

Comment:
doesn't have to be in this PR, but we should avoid the double lookup of find-and-insert by using the return value of `insert`.

Reply:
The find and insert are called on different sets (ancestors_`b`.find(a) and ancestors_`a`.insert(a)).